phone-book/src/edit-user.js

```diff
@@ -0,0 +1,195 @@
+
+
+class EditUser {
+  constructor(app) {
+    this.title = 'Add User'
+    this.app = app;
+    this.appContainer = app.appContainer;
+  }
+
+  initEvents(){
+    // this.createPlaceholders();
+    this.initSaveEvent();
+    if(this.app.state.userId){
+      this.initDeleteEvent();
+      this.renderUserData();
+    }
+  }
+
+  setCaretToEnd(node) {
+    let range = document.createRange();
+    let sel = window.getSelection();
+    range.setStart(node, 1);
+    range.collapse(true);
+    sel.removeAllRanges();
+    sel.addRange(range);
+  }
+  renderUserData(){
+    const contactObj = this.app.pages.contacts.pageObject;
+    const userId = this.app.state.userId;
+    this.getUser();
+  }
+
+  getUser(){
+    const userId = this.app.state.userId;
+    const serverAPI = this.app.serverAPI;
+    const loadingUser = serverAPI.getUser(userId);
+    loadingUser.then( response => {
+      return response.json();
+    })
+    .then(user => {
+      const fullNameField = this.appContainer.querySelector('span[data-name="fullName"]');
+      fullNameField.textContent = user.fullName;
+
+      const emailField = this.appContainer.querySelector('span[data-name="email"]');
+      emailField.textContent = user.email;
+
+      const phoneField = this.appContainer.querySelector('span[data-name="phone"]');
+      phoneField.textContent = user.phone;
+    })
+    .catch( err => {
+      console.error(err.message);
+    });
+
+  }
+
+  updateUser(data){
+    const serverAPI = this.app.serverAPI;
+    const updatingUser = serverAPI.updateUser(data);
+    updatingUser.then( response => {
+     this.app.pages.contacts.pageObject.users = [];
+      const newPage = 'contacts';
+      const newState = { activePage : newPage };
+      this.app.updateState(newState);          
+      this.app.router.gotToPage(this.app.state);
+      this.app.changePageToActive();
+    })
+    .catch( err => {
+      console.error(err.message);
+    });    
+  }
+
+  addUserToServer(data){
+    const serverAPI = this.app.serverAPI;
+    const addingUser = serverAPI.addUser(data);
+    addingUser.then( response => {
+      return response.json();
+    })
+    .then(addedUser => {
+        this.app.pages.contacts.pageObject.users = [];
+        const newPage = 'user';
+        const newUser = addedUser._id;
+        const newState = { activePage : newPage };
+        if(newUser){
+          newState.userId = newUser;
+        }
+        this.app.updateState(newState);          
+        this.app.router.gotToPage(this.app.state);
+        this.app.changePageToActive();
+    })
+    .catch( err => {
+      console.error(err.message);
+    });    
+  }
+  deleteUser(id){
+    const serverAPI = this.app.serverAPI;
+    const deletingUser = serverAPI.deleteUser(id);
+    deletingUser.then( response => {
+      this.app.pages.contacts.pageObject.users = [];
```

Comment:
it looks like really mess.

I believed pageObject is required overhead to structure, but it just makes the code more confusing.
If you want to store something you have to move its state and work with it directly.

If you want to update state - make 1 method where all updates would happens.

You could desing your app state withholding in mind 1 idea "I need to log my new state in console every time something changing"